cas-server-core/src/main/java/org/jasig/cas/CentralAuthenticationServiceImpl.java

```diff
@@ -284,11 +285,21 @@ public ServiceTicket grantServiceTicket(
                     uniqueTicketIdGenKey, service.getId());
             throw new UnauthorizedSsoServiceException();
         }
-        
+
+        final List<Authentication> authentications = ticketGrantingTicket.getChainedAuthentications();
+        final Principal principal = authentications.get(authentications.size() - 1).getPrincipal();
+
+        final Map<String, Object> principalAttrs = registeredService.getAttributeReleasePolicy().getAttributes(principal);
+        if (!registeredService.getAuthorizationStrategy().isServiceAccessAuthorizedForPrincipal(principalAttrs, service)) {
+            logger.warn("ServiceManagement: Cannot grant service ticket because Service [{}] is not authorized for use by [{}].",
+                    service.getId(), principal);
+            throw new UnauthorizedServiceForPrincipalException();

```

Comment:
yes, it should be hopefully. the message that is bundled says that the user specifically is unauthorized to access the application and that some form of contacting the helpdesk is required. of course cas server logs tell a lot more.